pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -640,6 +790,61 @@ impl std::ops::Deref for TimelineKey {
     }
 }
 
+/// A totally ordered f32 subset we can use with sorting functions.
+mod finite_f32 {
+
+    #[derive(Clone, Copy, PartialEq)]
+    pub struct FiniteF32(f32);
+
+    impl std::fmt::Debug for FiniteF32 {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            std::fmt::Debug::fmt(&self.0, f)
+        }
+    }
+
+    impl std::fmt::Display for FiniteF32 {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            std::fmt::Display::fmt(&self.0, f)
+        }
+    }
+
+    impl std::cmp::Eq for FiniteF32 {}
+
+    impl std::cmp::PartialOrd for FiniteF32 {
+        fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl std::cmp::Ord for FiniteF32 {
+        fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+            self.0.partial_cmp(&other.0).expect("both are finite")
```

Comment:
This treats `-0.0` and `0.0` the same, violating the total order property. Better is to use [total_cmp](https://doc.rust-lang.org/nightly/std/primitive.f32.html#method.total_cmp). e.g. this:

```Rust
fn noisy_cmp(a: f32, b: f32) {
    println!("comparing {a}({:x}) with {b}({:x}): partial_cmp: {:?}, total_cmp: {:?}", a.to_bits(), b.to_bits(), a.partial_cmp(&b), a.total_cmp(&b));
}
fn main() {
    noisy_cmp(1.0, 0.0);
    noisy_cmp(0.0, -0.0);
    noisy_cmp(f32::NAN, f32::from_bits(0x7fc00123));
}
```

has the following output:

```
comparing 1(3f800000) with 0(0): partial_cmp: Some(Greater), total_cmp: Greater
comparing 0(0) with -0(80000000): partial_cmp: Some(Equal), total_cmp: Greater
comparing NaN(7fc00000) with NaN(7fc00123): partial_cmp: None, total_cmp: Less
```

Reply:
I wonder, given `total_cmp` existing, if we should just rename the struct to `TotalF32` and implement unconditional `From<f32>` for it.

Reply:
how does one create such FiniteF32 values for both -0.0 and 0.0 if they are different?

Reply:
One does, did not remember -0.0 == 0.0. Added further normalization in f0ee3e2e8

Reply:
in the call we had today I got convinced that one needs to divide by something.